Samples/Samples/ViewModel/ContactsViewModel.cs

```diff
@@ -91,29 +91,38 @@ async void OnGetContact()
             }
         }
 
-        void OnGetAllContact()
+        async Task OnGetAllContact()
+        {
+            if (DeviceInfo.Platform != DevicePlatform.iOS)
+                await Permissions.RequestAsync<Permissions.ContactsRead>();
+            GetAllContact();
+        }
+
+        void GetAllContact()
         {
             if (IsBusy)
                 return;
             IsBusy = true;
             try
             {
-                var collect = Contacts.GetAllAsync();
-
-                Task.Run(() =>
+                var tasks = Contacts.GetAllAsync();
+                if (tasks != null)
                 {
-                    foreach (var a in collect)
-                        MainThread.BeginInvokeOnMainThread(() => ContactsList.Add(a));
-                });
+                    foreach (var task in tasks)
+                    {
+                        Task.Run(async () =>
+                        {
+                             foreach (var contact in await task)
```

Comment:
@mattleibow @pictos 
I found a way for all platforms, but its use doesn't look very nice

Reply:
That is not very nice indeed...

What if instead of an IEnumerable result, we return some contacts result object?

Then this can have the actual work/paging/enumeration.

Thus, the contacts API really returns the iterator object. With think like hasmore and getnextasync and stuff like that?

Reply:
Maybe we can even have the iasyncenumerable on the playforms that support it... Everything bu UWP maybe

Reply:
we only need IAsyncEnumerable for uwp. But IAsyncEnumerable only works with C# version 8.0

Reply:
and I don't think giving up UWP is a good way

Reply:
Oh yeah. I mean can we have a class that conditionally implements the interface? I have used it in the past, but I am not sure if it supports an actual implementation.

Then we can either use await foreach on supporting platforms, and some nice GetNextAsync members on UWP.